File: frontend/lib/models/budget.dart
/// Budget model
class Budget {
  final String id;
  final String userId;
  final String category;
  final double monthlyLimit;
  final String financialYear;
  final DateTime createdAt;
  final DateTime updatedAt;
  final double? spent;
  final double? remaining;
  final double? percentageUsed;

  Budget({
    required this.id,
    required this.userId,
    required this.category,
    required this.monthlyLimit,
    required this.financialYear,
    required this.createdAt,
    required this.updatedAt,
    this.spent,
    this.remaining,
    this.percentageUsed,
  });

  factory Budget.fromJson(Map<String, dynamic> json) {
    return Budget(
      id: json['id'],
      userId: json['user_id'],
      category: json['category'],
      monthlyLimit: (json['monthly_limit'] as num).toDouble(),
      financialYear: json['financial_year'],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      spent: json['spent'] != null ? (json['spent'] as num).toDouble() : null,
      remaining: json['remaining'] != null ? (json['remaining'] as num).toDouble() : null,
      percentageUsed: json['percentage_used'] != null 
          ? (json['percentage_used'] as num).toDouble() 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'category': category,
      'monthly_limit': monthlyLimit,
      'financial_year': financialYear,
    };
  }
}

File: backend/app/schemas/budget.py
"""Budget schemas."""
from datetime import datetime
from decimal import Decimal
from typing import Optional
from pydantic import BaseModel, Field
from uuid import UUID


class BudgetBase(BaseModel):
    """Base budget schema."""
    category: str
    monthly_limit: Decimal = Field(..., gt=0)
    financial_year: str


class BudgetCreate(BudgetBase):
    """Budget creation schema."""
    pass


class BudgetUpdate(BaseModel):
    """Budget update schema."""
    category: Optional[str] = None
    monthly_limit: Optional[Decimal] = Field(None, gt=0)
    financial_year: Optional[str] = None


class BudgetResponse(BudgetBase):
    """Budget response schema."""
    id: UUID
    user_id: UUID
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class BudgetWithSpending(BudgetResponse):
    """Budget response with spending information."""
    spent: Decimal
    remaining: Decimal
    percentage_used: float
